source/extensions/quic_listeners/quiche/envoy_quic_server_stream.cc

```diff
@@ -204,35 +211,43 @@ void EnvoyQuicServerStream::OnBodyAvailable() {
     return;
   }
 
-  if (!quic::VersionUsesHttp3(transport_version()) && !FinishedReadingTrailers()) {
-    // For Google QUIC implementation, trailers may arrived earlier and wait to
-    // be consumed after reading all the body. Consume it here.
-    // IETF QUIC shouldn't reach here because trailers are sent on same stream.
-    request_decoder_->decodeTrailers(
-        spdyHeaderBlockToEnvoyHeaders<Http::RequestTrailerMapImpl>(received_trailers()));
-    MarkTrailersConsumed();
-  }
+  // Trailers may arrived earlier and wait to be consumed after reading all the body. Consume it
```

Comment:
Optional, do you think we should refactor the API here to not pass trailers on until the body has been consumed?  I can't think of any gains to calling OnTrailingHeadersComplete until the data is consumed

Reply:
There is a QUICHE bug filed for this. It makes no sense for QUICHE to deliver trailers out of order, especially in IETF QUIC where trailers are sent on the data stream.